### ISSUE TITLE
Unable to replicate trained Sli-Prompts from Evaluation Code

### ISSUE THREAD

**Issue by EthanLatimerGB** [iamwangyabin/SPrompts_eval#5]
I was attempting to evaluate my Sli-Prompts model, trained on CDDB-Hard, with this repositories evaluation code. I could not achieve the results reported in your paper. 

# My training method

## Setup

Only the seed was changed in the `cddb-slip.json` file. 
```json
{
    "prefix": "reproduce",
    "dataset": "cddb",
    "task_name": ["gaugan", "biggan", "wild", "whichfaceisreal", "san"],
    "data_path": "/home/wangyabin/workspace/datasets/DeepFake_Data/CL_data/",
    "multiclass": [0, 0, 0, 0, 0],
    "class_order": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
    "total_sessions" : 7,
    "memory_size": 0,
    "memory_per_class": 0,
    "fixed_memory": true,
    "shuffle": false,
    "init_cls": 2,
    "increment": 2,
    "model_name": "sprompts",
    "net_type": "slip",
    "embd_dim" : 768,
    "prompt_length" : 10,
    "device": ["0","1"],
    "seed": [1993, 819023, 180238, 132341, 10140],
    "EPSILON" : 1e-8,
    "init_epoch" : 20,
    "init_lr" : 0.001,
    "init_milestones" : [20,30,40],
    "init_lr_decay" : 0.1,
    "init_weight_decay" : 0.0005,
    "epochs" : 50,
    "lrate" : 0.01,
    "milestones" : [20, 30],
    "lrate_decay" : 0.1,
    "batch_size" : 128,
    "weight_decay" : 2e-4,
    "num_workers" : 16
}
```
A few changes were needed to make the evaluation code work for the trained model, but I dont think it should directly affect the results.

## Results
During training I was receiving results like this in the logs, keep in mind I'm almost certain these are results from the **test** dataset after being trained:
```
2025-03-11 00:28:54,245 [trainer.py] => CNN: {'total': 99.75, '00-01': 99.75, 'old': 0, 'new': 99.75}
2025-03-11 01:15:48,296 [trainer.py] => CNN: {'total': 99.68, '00-01': 99.8, '02-03': 99.38, 'old': 99.8, 'new': 99.38}
2025-03-11 02:53:08,577 [trainer.py] => CNN: {'total': 93.15, '00-01': 99.8, '02-03': 99.25, '04-05': 84.34, 'old': 99.64, 'new': 84.34}
2025-03-11 03:47:39,909 [trainer.py] => CNN: {'total': 93.43, '00-01': 99.8, '02-03': 99.12, '04-05': 84.39, '06-07': 96.75, 'old': 93.15, 'new': 96.75}
2025-03-10 23:54:25,108 [trainer.py] => CNN: {'total': 91.82, '00-01': 99.0, '02-03': 97.38, '04-05': 83.52, '06-07': 94.5, '08-09': 61.11, 'old': 92.34, 'new': 61.11}
```
Whereas the evaluation code, picking the trained model, I achieve equally bad results for all the snapshots:
```bash
#!/bin/bash

PATH_LOCATION=/lyceum/el3g21/el3g21/S-Prompts/logs/reproduce_10140_sprompts_slip_cddb_2_2_2025-03-11-13:29:28

python eval.py --resume $PATH_LOCATION/task_0.pth --dataroot /scratch/el3g21/CDDB/ --datatype deepfake &>> eval.txt
python eval.py --resume $PATH_LOCATION/task_1.pth --dataroot /scratch/el3g21/CDDB/ --datatype deepfake &>> eval.txt
python eval.py --resume $PATH_LOCATION/task_2.pth --dataroot /scratch/el3g21/CDDB/ --datatype deepfake &>> eval.txt
python eval.py --resume $PATH_LOCATION/task_3.pth --dataroot /scratch/el3g21/CDDB/ --datatype deepfake &>> eval.txt
python eval.py --resume $PATH_LOCATION/task_4.pth --dataroot /scratch/el3g21/CDDB/ --datatype deepfake &>> eval.txt                                                                         
```
Results were:
```
1.0
{'total': 74.89, '00-01': 99.75, '02-03': 94.12, '04-05': 49.15, '06-07': 50.5, '08-09': 50.0, 'task_wise': 68.704}
0.43003923033812813
{'total': 74.89, '00-01': 99.75, '02-03': 94.12, '04-05': 49.15, '06-07': 50.5, '08-09': 50.0, 'task_wise': 68.704}
0.4294787969362974
{'total': 74.89, '00-01': 99.75, '02-03': 94.12, '04-05': 49.15, '06-07': 50.5, '08-09': 50.0, 'task_wise': 68.704}
0.38464412478983745
{'total': 74.89, '00-01': 99.75, '02-03': 94.12, '04-05': 49.15, '06-07': 50.5, '08-09': 50.0, 'task_wise': 68.704}
0.3644685223239305
{'total': 74.89, '00-01': 99.75, '02-03': 94.12, '04-05': 49.15, '06-07': 50.5, '08-09': 50.0, 'task_wise': 68.704}
``` 
I've tried investigating why this is happening, but have not reached any sort of conclusion. I've checked if the cluster centers were changing for each snapshot. The reason for this was that I was attempting to see if maybe the model kept saving the old model during the training process, which explained only good performance on the first 2 datasets, but no. 
```
PATH 0:
Selection Statistics:
Selection 0: 5353 times

PATH 1:
Selection Statistics:
Selection 0: 2302 times
Selection 1: 3051 times

PATH 2:
Selection Statistics:
Selection 0: 2299 times
Selection 1: 979 times
Selection 2: 2075 times

...
```
# Similar results to the pre-trained model
Running the evaluation code on your provided pre-trained model was giving me similar results as-well:
```
0.34933682047450026
{'total': 73.98, '00-01': 99.85, '02-03': 87.75, '04-05': 49.15, '06-07': 50.5, '08-09': 50.0, 'task_wise': 67.45}
```

It's quite strange behaviour, would be interested in finding out if you are able to help me achieve the same results as you in the evaluation code.

**Comment by iamwangyabin** [comment]
Hello Thanks for reporting this issue. 
SPrompts is an incremental learning model designed for binary classification tasks (real/fake), but it handles this in a specific way that might not be immediately obvious:

While each individual task is a binary classification problem (real/fake), the model architecture progressively adds classification heads as new tasks are introduced. The model starts with 2 classes, then grows to 4, 6, 8, and finally 10 classes as you progress through tasks

So, the evaluation code needs to account for this architecture. Process the outputs using a modulo operation (%2) to convert the multi-class outputs back to binary decisions (real/fake). For example, if you have logits for classes 0-9, you would treat even indices (0,2,4,6,8) as one class and odd indices (1,3,5,7,9) as the other class.

I suggest you try the following troubleshooting steps:
Print all outputs from the classification head for a few test samples to see the actual output
Check that weights are correctly loaded for each task snapshot

Looking at your training logs, I can see results are actually within the expected range reported in paper, your training was successful.

**Comment by iamwangyabin** [comment]
Debugging this model doesn't require highperformance GPU 
I recommend testing with a sample locally first before deploying to HPC for comprehensive testing.
